### PR TITLE
X509_check_issued: check that signature algo matches signing key algo

### LINKED ISSUE #7899: issuer certificate may found unexpected and result verify failed
CA have signed certificate with two different public key algorithm, both CA have loaded in verify store, If the different was load before same  public key algorithm,the wrong CA will found to build the certificate chain,and verification will failed.

Version: 1.1.1
build: self build
Configure certificate (issuer and subject is same and no akid): 
Root CA(RSA) > CA(RSA) > EE(RSA)
Root CA(DSA) > CA(DSA) > EE(DSA)
Verify mode: Client verify Server
Client:  
Step 1: Init setup
Step 2: Load Root CA(DSA) and CA(DSA)
Step 3: Load Root CA(RSA) and CA(RSA)

Server:
Step 1: Init setup
Step2: Load EE(RSA) and key
Step 3:Load CA(RSA)

Build certificate chain:   Root CA(DSA) - CA(DSA) - EE(RSA) 

The verification with such certificate chain will  failed.

### PR DESCRIPTION
**Pull request by levitte**
This implements 3.5.18 "Consistent Public Key and Signature Algorithms"
from RFC 4158 "Internet X.509 Public Key Infrastructure: Certification
Path Building"

Ref: https://tools.ietf.org/html/rfc4158#section-3.5.18

Fixes #7899

### CODE REVIEW

**Comment by levitte** [comment]
I finally found the right way to do this.  Much simpler than I first anticipated.

This is not WIP any more. 

@MagicC1, please have a look and see if this solves your probem.

**Comment by levitte** [comment]
Kinks worked out...  it's time for a review.

File: crypto/x509v3/v3_purp.c [hunk 1]
```diff
@@ -765,7 +765,8 @@ static int no_check(const X509_PURPOSE *xp, const X509 *x, int ca)
  * These are:
  * 1. Check issuer_name(subject) == subject_name(issuer)
  * 2. If akid(subject) exists check it matches issuer
- * 3. If key_usage(issuer) exists check it supports certificate signing
+ * 3. Check that signing algorithm matches signature algorithm
+ * 4. If key_usage(issuer) exists check it supports certificate signing
```
Comment (mspncp) [inline crypto/x509v3/v3_purp.c]: "If ... exists check it supports ..." -> "If ... exists, check whether/if/that it supports..."   (note the added ",").

Applies also to  2.

File: crypto/x509v3/v3_purp.c [hunk 2]
```diff
@@ -785,6 +786,24 @@ int X509_check_issued(X509 *issuer, X509 *subject)
             return ret;
     }
 
+    {
+        /*
+         * Check if the subject signature algorithm matches the issue PUBKEY
```
Comment (mspncp) [inline crypto/x509v3/v3_purp.c]: issue -> issuer's

File: include/openssl/x509_vfy.h [hunk 3]
```diff
@@ -184,6 +184,10 @@ void X509_STORE_CTX_set_depth(X509_STORE_CTX *ctx, int depth);
 # define         X509_V_ERR_OCSP_VERIFY_NEEDED                   73  /* Need OCSP verification */
 # define         X509_V_ERR_OCSP_VERIFY_FAILED                   74  /* Couldn't verify cert through OCSP */
 # define         X509_V_ERR_OCSP_CERT_UNKNOWN                    75  /* Certificate wasn't recognized by the OCSP responder */
+/* More 'informational' when looking for issuer cert */
+# define         X509_V_ERR_SIGNATURE_ALGORITHM_MISMATCH         76
+/* When the issuer certificate doesn't have a public key */
+# define         X509_V_ERR_NO_ISSUER_PUBLIC_KEY                 77
```
Comment (mspncp) [inline include/openssl/x509_vfy.h]: Why did you choose a different alignment? The `X509_V_ERR_OCSP_CERT_UNKNOWN` line already has 127 columns, so it does not make sense to insist on the 80 column rule here, unless you reformat all lines. But the way the existing comments are is more grep friendly, which is IMHO a legal reason for ignoring the 80 column rule here.

**Review by mspncp — CHANGES_REQUESTED** [review]

File: include/openssl/x509_vfy.h (hunk 3 above)
Comment (levitte) [reply include/openssl/x509_vfy.h]: I didn't pay attention to the previous ones.  Will fix

File: crypto/x509v3/v3_purp.c (hunk 2 above)
Comment (levitte) [reply crypto/x509v3/v3_purp.c]: Will fix

File: crypto/x509v3/v3_purp.c (hunk 1 above)
Comment (levitte) [reply crypto/x509v3/v3_purp.c]: Will fix

File: include/openssl/x509_vfy.h [hunk 4]
```diff
@@ -185,6 +185,10 @@ void X509_STORE_CTX_set_depth(X509_STORE_CTX *ctx, int depth);
 # define         X509_V_ERR_OCSP_VERIFY_FAILED                   74  /* Couldn't verify cert through OCSP */
 # define         X509_V_ERR_OCSP_CERT_UNKNOWN                    75  /* Certificate wasn't recognized by the OCSP responder */
 
+# define         X509_V_ERR_SIGNATURE_ALGORITHM_MISMATCH         76
+# define         X509_V_ERR_NO_ISSUER_PUBLIC_KEY                 77
```
[the remark this answers is not in the document]
Comment (mspncp) [reply include/openssl/x509_vfy.h]: Forget my last comment. I had a look at the complete file and only three entries have comments. This was not visible in the diff.

**Review by mspncp — APPROVED** [review]

File: include/openssl/x509_vfy.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (mspncp) [reply include/openssl/x509_vfy.h]: Maybe you could delete the empty line 187 while merging?

[the remark this answers is not in the document]
Comment (levitte) [reply include/openssl/x509_vfy.h]: Yup.  If you look, you'll notice that almost all of the others lack comments, so...  besides, there are text descriptions in `crypto/x509/x509_txt.c`

**Comment by levitte** [comment]
Merged.

ffd2df135a5d9f6d2627bd125f362298430fdc06 X509_check_issued: check that signature algo matches signing key algo

**Pull request closed without merging**